api/adapters/hydroshare.py

```diff
@@ -0,0 +1,260 @@
+import requests
+from starlette import status
+from datetime import datetime
+from typing import List, Optional, Union
+from pydantic import BaseModel, EmailStr, HttpUrl
+from api.adapters.base import AbstractRepositoryMetadataAdapter, AbstractRepositoryRequestHandler
+from api.exceptions import RepositoryException
+from api.models import schema
+from api.models.catalog import DatasetMetadataDOC
+
+
+class _HydroshareRequestHandler(AbstractRepositoryRequestHandler):
+
+    def get_metadata(self, record_id: str):
+        hs_meta_url = self.settings.hydroshare_meta_read_url % record_id
+        hs_file_url = self.settings.hydroshare_file_read_url % record_id
+
+        def make_request(url, file_list=False):
+            response = requests.get(url)
+            if response.status_code != status.HTTP_200_OK:
+                raise RepositoryException(status_code=response.status_code, detail=response.text)
+            if not file_list:
+                return response.json()
+
+            content_files = []
+            content_files.extend(response.json()["results"])
+            # check if there are more results to fetch - by default, 100 files are returned from HydroShare
+            while response.json()["next"]:
+                response = requests.get(response.json()["next"])
+                if response.status_code != status.HTTP_200_OK:
+                    raise RepositoryException(status_code=response.status_code, detail=response.text)
+                content_files.extend(response.json()["results"])
+            return content_files
+
+        metadata = make_request(hs_meta_url)
+        files_metadata = make_request(hs_file_url, file_list=True)
+        metadata["content_files"] = files_metadata
+        return metadata
+
+
+class HydroshareMetadataAdapter(AbstractRepositoryMetadataAdapter):
+    repo_api_handler = _HydroshareRequestHandler()
+
+    @staticmethod
+    def to_catalog_record(metadata: dict) -> DatasetMetadataDOC:
+        """Converts hydroshare resource metadata to a catalog dataset record"""
+        hs_metadata_model = _HydroshareResourceMetadata(**metadata)
+        return hs_metadata_model.to_catalog_dataset()
+
+    @staticmethod
+    def to_repository_record(catalog_record: DatasetMetadataDOC):
+        """Converts dataset catalog record to hydroshare resource metadata"""
+        raise NotImplementedError
+
+
+class _HydroshareResourceMetadata(BaseModel):
+
+    class Creator(BaseModel):
```

Comment:
Classes defined inside a class make it hard to read the parent class.  Could you move these classes into their own file?  I also suggest moving the metadata models into their own classes so they can be reused without coupling them to the adapters.

Reply:
Fair point about reuse.  Perhaps this is more a matter of taste but I do think embedded class definitions make it more difficult to read the parent class.  A user has to scroll through 150 lines of class definitions to even get to the properties of the parent class.